Paper Implementation Project_785: Qauntum walks via efficient block encoding circuits for structured stochastic matrices

virajd98 opened the pull request:
### PR Description

This PR implements key components of [this paper ](https://arxiv.org/abs/2203.10236) and is a part of [this issue](https://github.com/Classiq/classiq-library/issues/785).

**Deliverable includes a detailed Jupyter notebook containing:**
- Circuit implementations of block encoding of adjacency matrix of an extended binary tree graph and a exact Hermitian block encoding of a stochastic banded circulant matrix
- Performance benchmarks and resource analysis of this technique vs the already existing LCU technique available [here](https://docs.classiq.io/latest/classiq_101/quantum_primitives_with_classiq/linear_combination_of_unitaries/?h=linear+com) or [here](https://docs.classiq.io/latest/explore/tutorials/basic_tutorials/hamiltonian_simulation/hamiltonian_simulation_with_block_encoding/hamiltonian_simulation_with_block_encoding/?h=block+encoding).
- Application of this for quantum walks


### Some notes

- [ ] Please make sure that you placed the files in an appropriate folder
- [ ] And that the files have indicative names.

- [ ] Please note that Classiq runs automatic code linting, which may minorly alter some files.
  - [ ] If you're familiar with `pre-commit`, you may run `pre-commit install`, and then at each commit, your files will be altered in a similar way

orsa-classiq commented:
@virajd98 Amazing job!
I started to review it, put comments until the random walk part. You can start to fix them
Notice the submission guidelines here: https://github.com/Classiq/classiq-library/blob/main/community/paper_implementation_project/index.md
(also put you files in a directory under this path)

virajd98 commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772220950'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772220978'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Thanks, I missed the O_A. </p><p> </p><p>The dimensions seem fine to me. </p><p> </p><p>It's a general technique to block encode given the oracles can be efficiently constructed. Yes, in the first definition<code> a</code> is the number of auxiliaries needed for block encoding.</p><p> </p><p>Yes, l2 norm of all matrices considered in the notebook is&nbsp;≤1.&nbsp;I have added this point, thanks</p><p> </p><p>Done. </p><p> </p><p> </p><p> </p><p> </p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221006'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>You can think of the measured register as part of the block. More specifically, the outcome here indicates if M2 operation should be kept or discarded. If its |0&gt;, then we keep it (as we usually do, block ==0) and this gives the complete action of given non unitary matrix on any input state. </p><p> </p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221048'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221073'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Thanks, this works for this case.</p><p> </p><p>But please note, in section 2 of my notebook, where I again call <code>L_shift </code> the code rewritten with += or -= doesn't seem to give the right answer? I don't understand why this is the case, so I kept the same functions in section 2. Please check if you can see why this issue arises. </p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221104'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221135'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done. This was useful, as so many lines now got reduced</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221167'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221205'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221326'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done. </p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221350'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>I have explained what a marked vertex means- refer the para- In a classical random walk,&nbsp;....</p><p> </p><p>Yes, the explanation for how P is constructed is not given. It's more like, consider this P and now we study its properties and when we do this we see that random walks associated with this P have specific convergence properties. Example, for the P of the non-marked case the random walk always results in a state (or position of walker) as equal superposition of all positions and so on...</p><p> </p><p>Stochastic matrix (P) determines the graph. So as per my understanding, it's redundant to define both P and then say what's the graph. </p><p> </p><p>Ok, in the full notebook, N=2^n, I'll mention this. Thanks</p><p> </p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772655010'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772655119'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Done, I have used execution session as context manager, now the runtime has reduced a lot. Thanks</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772655160'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>Thanks, I have included the final expression for p(|0&gt;) and you can see it has a term with Chebyshev polynomial which has oscillatory behavior. </p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772655252'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>The first one is for just N= 4 vertices and you can see that blue curve doesn't come close to the intersection of the two dotted lines. But if you increase N, you can observe that it gets closer. So the result is valid in large N regime. So two plots to justify this was my idea</p><p> </p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772655298'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>As per my understanding of the provided definition, c(j,l) (row index) is only defined if a specific j (column index) and l (sparsity index) correspond to a non-zero value at the row index. For the example you gave, l=0 and j&gt;=2^(n-1), such values don't correspond to any row index that has a non-zero value (by definition or construction)</p><p> </p><p>Regarding invertibility, both (1, 4) and (1, 5) refer to column 1 and as per definition of 'l' (i.e 3&lt;l&lt;8 correspond to sparsity indices of the diagonal values), both (1, 4) and (1, 5) should point to the row index of the diagonal at j=1, which is 1. Hence (1, 4) and (1, 5) maps to (1, 1) matches the logic of the construction. I'm not sure why invertibility of c(j,l) is a necessity?</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221029'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
<p>I incorporated the within_apply suggestion. I don't think there is reflection about zero state in the code.</p>

---

 <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772655079'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

virajd98 commented:
@orsa-classiq  In the updated pull request, I have addressed all feedbacks and the submission guidelines have been followed. Let me know if anything more is needed.

orsa-classiq commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221029) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-02T11:08:37Z_ ----------------------------------------------------------------
> 
> in the equation for O_c it seemed to be inplace operation, so (j, l) -> (j, c(j, l)) should be invertible. But this map seems not to be invertible (i.e both (1, 4) and (1, 5) map to (1, 1)). How is it solved?
> 
> 
> It is not complete - what is the definition for l=0 and j>=2^(n-1)?
> 
> 
> 
> 
> 
> _virajd98 commented on 2025-04-04T17:22:24Z_ ----------------------------------------------------------------
> 
> As per my understanding of the provided definition, c(j,l) (row index) is only defined if a specific j (column index) and l (sparsity index) correspond to a non-zero value at the row index. For the example you gave, l=0 and j>=2^(n-1), such values don't correspond to any row index that has a non-zero value (by definition or construction)
> 
> 
> Regarding invertibility, both (1, 4) and (1, 5) refer to column 1 and as per definition of 'l' (i.e 3<l<8 correspond to sparsity indices of the diagonal values), both (1, 4) and (1, 5) should point to the row index of the diagonal at j=1, which is 1. Hence (1, 4) and (1, 5) maps to (1, 1) matches the logic of the construction. I'm not sure why invertibility of c(j,l) is a necessity?

The point is that a unitary cannot map |1, 4> -> |1, 1> and also |1, 5> -> |1, 1>. That is because it is not invertible, hence not unitary. If the operation you do is out of place, say |1, 4, 0> -> |1, 4, 1> then it is possible. It is not clear from the text

orsa-classiq commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221228) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-02T11:08:42Z_ ----------------------------------------------------------------
> 
> Line #3.    def unitary_K(block: QArray):
> don't you miss the a loop for the needed CX gates according to the diagram? or is it only for 3 qubits case? if so need to specify
> 
> 
> _virajd98 commented on 2025-04-02T16:34:21Z_ ----------------------------------------------------------------
> 
> This is only for 3 qubit case. From the paper it was not evident to me how `unitary_K` can be generalized to a n-qubit case. Can you see if you can understand the generalization? That would be useful for the implementation.
> 
> 
> I'll specify for now that this is for 3 qubits and generalizations can be explored

It seems that the diagram that you provided is just what you need - 2 rotations, CX on every 2 qubits except the first 2, then a toffoli

orsa-classiq commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772655079) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-02T13:58:37Z_ ----------------------------------------------------------------
> 
> Line #7.        control(block==0, lambda: X(anc))   
> you can also do: `anc ^= (block == 0)` (in this case use `within_apply` because of the compute-uncompute sequence)
> 
> also the entire first part is just `reflect_about_zero` I think, you can use the function.
> 
> 
> _virajd98 commented on 2025-04-04T17:37:49Z_ ----------------------------------------------------------------
> 
> I incorporated the within_apply suggestion. I don't think there is reflection about zero state in the code.

https://docs.classiq.io/latest/qmod-reference/api-reference/functions/open_library/grover/?h=reflect#classiq.open_library.functions.grover.reflect_about_zero

virajd98 commented:
> > View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221029) on ReviewNB
> > _orsa-classiq commented on 2025-04-02T11:08:37Z_ ----------------------------------------------------------------
> > in the equation for O_c it seemed to be inplace operation, so (j, l) -> (j, c(j, l)) should be invertible. But this map seems not to be invertible (i.e both (1, 4) and (1, 5) map to (1, 1)). How is it solved?
> > It is not complete - what is the definition for l=0 and j>=2^(n-1)?
> > _virajd98 commented on 2025-04-04T17:22:24Z_ ----------------------------------------------------------------
> > As per my understanding of the provided definition, c(j,l) (row index) is only defined if a specific j (column index) and l (sparsity index) correspond to a non-zero value at the row index. For the example you gave, l=0 and j>=2^(n-1), such values don't correspond to any row index that has a non-zero value (by definition or construction)
> > Regarding invertibility, both (1, 4) and (1, 5) refer to column 1 and as per definition of 'l' (i.e 3<l<8 correspond to sparsity indices of the diagonal values), both (1, 4) and (1, 5) should point to the row index of the diagonal at j=1, which is 1. Hence (1, 4) and (1, 5) maps to (1, 1) matches the logic of the construction. I'm not sure why invertibility of c(j,l) is a necessity?
> 
> The point is that a unitary cannot map |1, 4> -> |1, 1> and also |1, 5> -> |1, 1>. That is because it is not invertible, hence not unitary. If the operation you do is out of place, say |1, 4, 0> -> |1, 4, 1> then it is possible. It is not clear from the text

The unitary does not  map |1, 4> -> |1, 1> and |1, 5> -> |1, 1>. Rather it is  |1, 4> -> |1, 4> and |1, 5> -> |1, 5>. Notice the definition of O_c is not |j,l> to |j, c(j,l)> rather |j,l> to |c(j,l), l>. There is no need for c(j,l) to be invertible for O_c to be unitary here because of the way 0_c is defined. This operation is inplace on register 'j' . 
Regarding your earlier question on the definition of c(j,l) not being complete, such inputs are not relevant here and using ancilla we discard them. For example- You can see that in all cases where the inputs to c(j,l) isn't in the definition (say j=6, l=0 for the case n=3) the M_2 circuit will switch the ancilla to |1> and we discard this in post processing. So in the definition for c(j,l) only the relevant I/O are written because c(j,l) corresponds to something physical in the problem.

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889958) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-14T08:37:16Z_ ----------------------------------------------------------------
> 
> Line #12.    def main(k:CReal, anc: Output[QNum], block: Output[QNum], state: Output[QNum]):
> this block contains almost exactly the same code twice, wrap it into a function to reduce code duplication

I did not understand. Where is the duplication inside `main(k:CReal, anc: Output[QNum], block: Output[QNum], state: Output[QNum]):`?

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889915) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-14T08:37:15Z_ ----------------------------------------------------------------
> 
> Line #13.        apply_to_all(H, state)      
> you can also use `hadamard_transform`

Done.

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889881) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-14T08:37:14Z_ ----------------------------------------------------------------
> 
> Line #11.    print("The implementation is correct")
> remove this. In all places

Done

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889818) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-14T08:37:12Z_ ----------------------------------------------------------------
> 
> Line #34.                              control(l_array[2]==1, lambda: control(j_array[int(np.log2(n))-1]==0, lambda: RY(theta_1, data))) , 
> in all of these double controls you can use a single statement. such as (q[2]==0) & (q[1] == 1)

Thanks, I did not know that this was allowed. Done

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889915) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-14T08:37:15Z_ ----------------------------------------------------------------
> 
> Line #13.        apply_to_all(H, state)      
> you can also use `hadamard_transform`

this does NOT work in QNum types

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889718) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-14T08:37:10Z_ ----------------------------------------------------------------
> 
> the \ket latex is not rendering well on our documentation, better to replace it with |x\rangle
> 
> how many qubits do you need for this block encoding? mention it

Done.

virajd98 commented:
> > View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772655010) on ReviewNB
> > _orsa-classiq commented on 2025-04-02T13:58:36Z_ ----------------------------------------------------------------
> > "If there is a marked vertex in the graph, its corresponding probability ... will have a significantly higher magnitude, revealing its location." - You didn't explain what is a marked vertex and how do you build the P matrix from the graph and the marked vertices
> > cases 1 and 2: you didn't specify what is the underlying graph (I guess the fully connected graph?)
> > you didn't defined what is `n`.when you say P(|0^n>) you mean the prob to still be in the block?
> > _virajd98 commented on 2025-04-02T16:43:01Z_ ----------------------------------------------------------------
> > I have explained what a marked vertex means- refer the para- In a classical random walk, ....
> > Yes, the explanation for how P is constructed is not given. It's more like, consider this P and now we study its properties and when we do this we see that random walks associated with this P have specific convergence properties. Example, for the P of the non-marked case the random walk always results in a state (or position of walker) as equal superposition of all positions and so on...
> > Stochastic matrix (P) determines the graph. So as per my understanding, it's redundant to define both P and then say what's the graph.
> > Ok, in the full notebook, N=2^n, I'll mention this. Thanks
> 
> I looked for it and it is not explained. You used the term 'marked vertex' without explaining it.
> 
> Regarding the second point - it seems from your writing that every stochastic matrix looks like this, and it is not the case. If you want, you can say somehting like - 'take the stochastic matrix that represents the fully connected graph with uniform weights ... '

Done.

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889661) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-14T08:37:09Z_ ----------------------------------------------------------------
> 
> "The parameters ... represent the _subnormalization factor*- which adjusts for encoding matrices of any norm, and the *number of ancilla qubits_ used in the block-encoding scheme." - something weird here with the asterisks

Yes, there should be a 'respectively' written here. Done. The parameters $(\alpha, a)$  represent the *subnormalization factor* (which adjusts for encoding matrices of any norm), and the *number of ancilla qubits* used in the block-encoding scheme respectively.

virajd98 commented:
> > View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2772221228) on ReviewNB
> > _orsa-classiq commented on 2025-04-02T11:08:42Z_ ----------------------------------------------------------------
> > Line #3.    def unitary_K(block: QArray):
> > don't you miss the a loop for the needed CX gates according to the diagram? or is it only for 3 qubits case? if so need to specify
> > _virajd98 commented on 2025-04-02T16:34:21Z_ ----------------------------------------------------------------
> > This is only for 3 qubit case. From the paper it was not evident to me how `unitary_K` can be generalized to a n-qubit case. Can you see if you can understand the generalization? That would be useful for the implementation.
> > I'll specify for now that this is for 3 qubits and generalizations can be explored
> 
> It seems that the diagram that you provided is just what you need - 2 rotations, CX on every 2 qubits except the first 2, then a toffoli

This does not work. I have tried to understand the generalization from the circuit pictures in the paper and based on this tried different implementations. None of these seem to work, you can uncomment the #try1 or #try2 comments in both `unitary_K` function as well as `C_shiftAdder_SWAP` in the recent updates to see. Let me know if I'm missing something on the generalization. For now its working only for 8*8.

orsa-classiq commented:
> > Almost there!
> > 
> > 1. read my comments along the notebook and directly here and fix
> > 2. add a representative qmod to the directory, look for examples in the `algorithms` directory. make sure to follow all contribution guidelines
> > 3. make sure you pass the CI tests (currently you don't pass them)
> 
> 1. Almost done, except for generalization of block encoding stochastic banded circulant matrix.
> 2. Done
> 3. I have no clue how to resolve this. Can you tell me what is causing the issue? I did everything as asked in the reviews.

regarding 3:
- you have a space ' ' in the directory name. It causes errors.
- Also install the pre-commit hooks and apply them on your files, it should fix the other errors
- you miss `<qmod_name>.metadata.json` files for the qmods 
- you miss a test file, and timeout records in the `timeouts.yaml` file

please read carefully the contribution guidelines, everything should be mentioned there

let us know when everything is ready for a check

virajd98 commented:
@orsa-classiq 
It seems to me that I have added everything as per the guidelines. I don't understand what more to do, the errors I see are not related to the files I worked with.

virajd98 commented:
@orsa-classiq I have now passed the test. It took around 8 minutes for my test case to run for in the 'Run Notebooks' test. I'm not sure if this is normal.

orsa-classiq commented:
> > View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889915) on ReviewNB
> > _orsa-classiq commented on 2025-04-14T08:37:15Z_ ----------------------------------------------------------------
> > Line #13.        apply_to_all(H, state)
> > you can also use `hadamard_transform`
> 
> this does NOT work for QNum types

it should work - there is an automatic casting of types when calling functions. Try and see it works, then replace in all places

orsa-classiq commented:
@virajd98 I've replied an added a few comments. Please fix soon and will be able to merge it

virajd98 commented:
> > def main(k:CReal, anc: Output[QNum], block: Output[QNum], state: Output[QNum]):
> 
> you have 2 main functions, synthesis and execution, where the only difference is calling `non_marked_discriminant` or `marked_discriminant`. Also the loop `or k in repetitions:` after is just the same for both of them. factor this logic to a (regular) function that can be used with either of the versions. Then the code will be shorter

Done

virajd98 commented on community/paper_implementation_project/Explicit_quantum_circuits_for_Block_encoding/extended_binary_tree_matrix_block_encoding.metadata.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "friendly_name": "extendedbinarytree_matrix_block_encoding",
```
Done

virajd98 commented:
> > > View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889915) on ReviewNB
> > > _orsa-classiq commented on 2025-04-14T08:37:15Z_ ----------------------------------------------------------------
> > > Line #13.        apply_to_all(H, state)
> > > you can also use `hadamard_transform`
> > 
> > 
> > this does NOT work for QNum types
> 
> it should work - there is an automatic casting of types when calling functions. Try and see it works, then replace in all places

![image](https://github.com/user-attachments/assets/d23d0022-4b48-4ef7-ad03-4994ee0c69aa)
![image](https://github.com/user-attachments/assets/2930e764-9596-4443-9ca3-eb2caa726c2c)

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2827314098) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-24T11:40:09Z_ ----------------------------------------------------------------
> 
> Line #14.    def main(k: CReal, anc: Output[QNum], block: Output[QNum], state: Output[QNum]):
> k should be a 'CInt'.

Done

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2827314031) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-24T11:40:07Z_ ----------------------------------------------------------------
> 
> Line #21.        # try 1
> after the first 2 lines, should be something like:
> 
> `repeat(block.len-3, lambda i: control(block[block.len-1], X(block[index+2]))`
> 
> `control(block[0:2], lambda: X(block[block.len-1]))`
> 
> anyway `n` in your code is not defined, I replaced it with block.len
> 
> If that doesn't work, then remove the comments ('try 1\2') and write that the function is defined only for 3 qubits. But I prefer that it will work

I did define 'n'. The notebook has to be run cell by cell, from top to bottom. 
![image](https://github.com/user-attachments/assets/95ec11f6-fcd0-4106-9796-433af82cc031)

virajd98 commented:
> > View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2827314031) on ReviewNB
> > _orsa-classiq commented on 2025-04-24T11:40:07Z_ ----------------------------------------------------------------
> > Line #21.        # try 1
> > after the first 2 lines, should be something like:
> > `repeat(block.len-3, lambda i: control(block[block.len-1], X(block[index+2]))`
> > `control(block[0:2], lambda: X(block[block.len-1]))`
> > anyway `n` in your code is not defined, I replaced it with block.len
> > If that doesn't work, then remove the comments ('try 1\2') and write that the function is defined only for 3 qubits. But I prefer that it will work
> 
> I did define 'n'. The notebook has to be run cell by cell, from top to bottom. ![image](https://private-user-images.githubusercontent.com/145847869/437333261-95ec11f6-fcd0-4106-9796-433af82cc031.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1NjYxMzUsIm5iZiI6MTc0NTU2NTgzNSwicGF0aCI6Ii8xNDU4NDc4NjkvNDM3MzMzMjYxLTk1ZWMxMWY2LWZjZDAtNDEwNi05Nzk2LTQzM2FmODJjYzAzMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNVQwNzIzNTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZDVmZDBiNGRmNmI0YWYzMzhkMmYxZGIxNjk2YTM2NmM5Mjg3ZDdlOWUwNjc5YmNiZjBhMDA5NzZhNjBiMjc2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.MeDBx9fXY0iymyt0xbhrZMVWQhPHPFeGwwEBMe-LcEQ)

Your code is exactly the same what I did in #try 2. It does not work. 
![image](https://github.com/user-attachments/assets/c13780f5-64cd-49e0-b745-7eaec49749ec)

virajd98 commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2827313998) on ReviewNB
> 
> _orsa-classiq commented on 2025-04-24T11:40:07Z_ ----------------------------------------------------------------
> 
> you still have \ket in the notebook. please remove all of them

Done

virajd98 commented:
@orsa-classiq  All comments have been addressed.

virajd98 commented:
> > > View / edit / reply to [this conversation](https://app.reviewnb.com/Classiq/classiq-library/pull/901/discussion/#comment-2800889915) on ReviewNB
> > > _orsa-classiq commented on 2025-04-14T08:37:15Z_ ----------------------------------------------------------------
> > > Line #13.        apply_to_all(H, state)
> > > you can also use `hadamard_transform`
> > 
> > 
> > this does NOT work for QNum types
> 
> it should work - there is an automatic casting of types when calling functions. Try and see it works, then replace in all places

Done. Thanks.

orsa-classiq commented:
Nice job @virajd98 !
